[WIP][RFC][Cart] Cart taxes scenarios

Zales0123 opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets |
| License       | MIT
| Doc PR        |

According to awesome work of @Arminek with behat scenarios [here](https://github.com/Sylius/Sylius/pull/3819) and upcoming changes with taxes [here](https://github.com/Sylius/Sylius/pull/3907), there are some new scenarios for dealing with taxes in cart.
Of course, they could be green after merging whole taxes-related stuff. They are not implemented yet, I'm just wondering about community feedback about theirs style, have I properly caught concept of new scenarios ;)

pjedrzejewski commented on features/shop/cart_with_different_taxes_for _different_zones.feature:
```diff
@@ -0,0 +1,48 @@
+    Background:
+        Given that store is operating on the France channel
+        And default currency is "EUR"
+        And tax rate "EU VAT" with 23% rate belongs to "Taxable Goods" category for "EU" zone
```
And store has "EU VAT" tax rate of 23% for "Taxable Goods" in "EU" zone
I'd use a different example here, Taxable Goods is too general, let's say it is food category or something like WDYT?

Zales0123 replied:
:+1: I was wondering about sth conciser, but didn't know how short can it be :smile: And :+1: for more specific category

pjedrzejewski commented on features/shop/cart_with_different_taxes_for _different_zones.feature:
```diff
@@ -0,0 +1,48 @@
+        And tax rate "No tax" with 0% rate belongs to "Taxable Goods" category for "Rest of World" zone
+        And tax rate "Low VAT" with 5% rate belongs to "Low-taxed Goods"
+        And there is user "john@example.com" identified by "password123"
+        And catalog has a product "PHP T-Shirt" priced at €100.00 with "Taxable Goods" as tax category
```
I'd split this into 2 steps. (we have fast Behat now :D)

``And store has a product "Bananas" priced at €2``
``And "Bananas" tax category is "Food"``

Something like that, this way we can build the product and have more flexibility, without too much detail in a single step.

pjedrzejewski commented on features/shop/cart_with_different_taxes_for _different_zones.feature:
```diff
@@ -0,0 +1,48 @@
+        And catalog has a product "Symfony Mug" priced at €50.00 with "Low-taxed Goods" as tax category
+        And store has free shipping method
+        And store allows paying offline
+        And I am logged in as "john@example.com"
```
We need to cover following cases:

* Completely anonymous Visitor
* Visitor who has entered his address in the checkout
* Customer who signed in and has no address
* Customer who signed in and has a default address
* Customer who signed in and filled his address in checkout

Later we would like to improve the detection for anonymous users and preselect their country etc. But let's start from these cases.

WDYT?

pjedrzejewski commented:
We should be more explicit about the zone, I'd be fine with steps like:

```
And there is "EU" zone containing all members of European Union
```

I know that first step is assuming we are operating a single country store, but in these scenarios we should be more explicit so something like:

```
Given that store is operating via single channel of sales
And it is shipping to France, Spain and Germany
And there is "EU" zone containing all members of European Union
```

I think our scenario will evolve and I am just throwing random ideas here, we will improve them step by step, but we need to get rid of this old, anti-BDD style. ;)

peteward commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        And catalog has a product "PHP T-Shirt" priced at €100.00
+        And "PHP T-Shirt" tax category is "Clothes"
+        And catalog has a product "Symfony Mug" priced at €50.00
+        And "Symfony T-Shirt" tax category is "Mugs"
```
I think it should be "Symfony Mug"

peteward commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        And store allows paying offline
+        And I am logged in as "john@example.com"
+
+    Scenario: Proper taxes before addressing
```
`Correct tax calculation before addressing`?

michalmarcinkowski replied:
`Displaying correct tax before specifying address`?

peteward commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+
+    Scenario: Proper taxes after specifying address
+        Given I added product "PHP T-Shirt" to the cart
+        And I proceed selecting "Offline" payment method and "Uzbekistan" as address country
```
`as shipping country`?

michalmarcinkowski replied:
`When ...`

peteward commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        Then my cart taxes should be "€23.00"
+        And my cart total should be "€123.00"
+
+    Scenario: Proper taxes after specifying address
```
`specifying shipping address`? since that's the one that dictates your taxes (not billing)

michalmarcinkowski replied:
`Displaying correct tax after specifying shipping address`

peteward commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        Given I added product "PHP T-Shirt" to the cart
+        And I added 2 products "Symfony Mug" to the cart
+        And I proceed selecting "Offline" payment method and "Uzbekistan" as address country
+        Then my cart taxes should be "€23.00"
```
how do you get any tax when shipping to Uzbekistan? You've only got non-zero tax rates configured for EU

peteward commented on features/shop/cart_with_multiple_tax_rates.feature:
```diff
@@ -0,0 +1,30 @@
+    Background:
+        Given that store is operating on the France channel
+        And default currency is "EUR"
+        And store has "EU VAT" tax rate of 23% for "Clothes"
```
is this feature not supposed to be zone-based? Might be worth explaining that somewhere if not.

Zales0123 replied:
My idea was, that if it's not zone specified in step it uses default zone (which definitely imply need for ``Default zone is "..."`` step :smile:)

peteward commented on features/shop/cart_with_multiple_tax_rates.feature:
```diff
@@ -0,0 +1,30 @@
+        And catalog has a product "PHP T-Shirt" priced at €100.00
+        And "PHP T-Shirt" tax category is "Clothes"
+        And catalog has a product "Symfony Mug" priced at 50.00
+        And "Symfony T-Shirt" tax category is "Mugs"
```
Mug

pamil commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        And store has "Low VAT" tax rate of 5% for "Mugs" in "EU" zone
+        And there is user "john@example.com" identified by "password123"
+        And catalog has a product "PHP T-Shirt" priced at €100.00
+        And "PHP T-Shirt" tax category is "Clothes"
```
"PHP T-Shirt" belongs to "Clothes" tax category?

pjedrzejewski replied:
@pamil @Zales0123 :+1:

pamil commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        And store has "EU VAT" tax rate of 23% for "Clothes" in "EU" zone
+        And store has "No tax" tax rate of 0% for "Clothes" in "Rest of World" zone
+        And store has "Low VAT" tax rate of 5% for "Mugs" in "EU" zone
+        And there is user "john@example.com" identified by "password123"
```
This one and the last step -> "I am logged in user"?

peteward commented on features/shop/cart_with_shipping_fee_and_product_taxes.feature:
```diff
@@ -0,0 +1,33 @@
+        And there is user "john@example.com" identified by "password123"
+        And store has "EU VAT" tax rate of 23% for "Clothes" in "EU" zone
+        And store has "Low tax" tax rate of 10% for "Clothes" in "Rest of world" zone
+        And catalog has a product "PHP T-Shirt" priced at €100.00
```
some prices / numbers in your gherkin are quoted and some aren't

peteward commented on features/shop/cart_with_shipping_fee_and_product_taxes.feature:
```diff
@@ -0,0 +1,33 @@
+        And catalog has a product "PHP T-Shirt" priced at €100.00
+        And "PHP T-Shirt" tax category is "Clothes"
+        And store has "DHL" shipping method with "€10.00" fee
+        And "DHL" tax category is "Clothes"
```
I suggest adding a tax category for "Shipping" to make this a bit clearer

peteward commented on features/shop/cart_with_shipping_fee_and_tax.feature:
```diff
@@ -0,0 +1,32 @@
+        And store has "Low tax" tax rate of 10% for "Clothes" in "Rest of world" zone
+        And catalog has a product "PHP T-Shirt" priced at €100.00
+        And store has "DHL" shipping method with "€10.00" fee
+        And "DHL" tax category is "Clothes"
```
again I think a Shipping category would be clearer

pamil commented on features/shop/cart_with_multiple_tax_rates.feature:
```diff
@@ -0,0 +1,30 @@
+        And catalog has a product "Symfony Mug" priced at 50.00
+        And "Symfony T-Shirt" tax category is "Mugs"
+        And store has free shipping method
+        And store allows paying offline
```
Do we need it here?

pamil commented on features/shop/cart_with_shipping_fee.feature:
```diff
@@ -0,0 +1,21 @@
+        And store has "DHL" shipping method with "€10.00" fee
+        And I am logged in as "john@example.com"
+
+
```
Doubled blank line :)

pamil commented on features/shop/cart_with_multiple_tax_rates.feature:
```diff
@@ -0,0 +1,30 @@
+        And there is user "john@example.com" identified by "password123"
+        And catalog has a product "PHP T-Shirt" priced at €100.00
+        And "PHP T-Shirt" tax category is "Clothes"
+        And catalog has a product "Symfony Mug" priced at 50.00
```
Missing "€"

michalmarcinkowski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+@ui-cart
+Feature: Cart with different tax rates for different time zones
```
`for different time zones` I suppose the zones have nothing to do with time :)

michalmarcinkowski replied:
Actually IMO the features description should be similar to:
`Apply taxes based on zone and tax rate`

pjedrzejewski replied:
:+1: for the second comment.

peteward commented:
Great to see these :+1: 

I suggest you add some scenarios that look at rounding and the calculation of tax at item / unit level as discussed by myself and @michalmarcinkowski in #3868.

Even if the default stance of Sylius doesn't agree with me, I think having scenarios that cover this are really important, and can be extended when multiple calculation strategies are implemented

michalmarcinkowski commented on features/shop/cart_with_multiple_tax_rates.feature:
```diff
@@ -0,0 +1,30 @@
+@ui-cart
+Feature: Cart with multiple tax rates
```
`Feature: Apply correct tax for items with different tax rates`

pamil commented:
From the implementation point of view, implementing step like this may be tough:
```gherkin
And "DHL" tax category is "Clothes" # DHL = Shipping method
And "Banana" tax category is "Fruits" # Banana = Product
```

pamil commented on features/shop/cart_with_shipping_fee.feature:
```diff
@@ -0,0 +1,21 @@
+        And default currency is "EUR"
+        And there is user "john@example.com" identified by "password123"
+        And catalog has a product "PHP T-Shirt" priced at €100.00
+        And store has "DHL" shipping method with "€10.00" fee
```
We should decide whether we write prices as `€0.50` or `"€3.00"`.

michalmarcinkowski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        And store has "No tax" tax rate of 0% for "Clothes" in "Rest of World" zone
+        And store has "Low VAT" tax rate of 5% for "Mugs" in "EU" zone
+        And there is user "john@example.com" identified by "password123"
+        And catalog has a product "PHP T-Shirt" priced at €100.00
```
`And there is a product "PHP T-Shirt" priced at €100.00 in the catalog`?

Zales0123 replied:
This step has been based on previously written payment scenario by @Arminek, but sure, I think it's more natural

michalmarcinkowski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        And catalog has a product "Symfony Mug" priced at €50.00
+        And "Symfony T-Shirt" tax category is "Mugs"
+        And store has free shipping method
+        And store allows paying offline
```
I would group all store configuration steps together, at the beginning.

michalmarcinkowski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+    Scenario: Proper taxes before addressing
+        Given I added product "PHP T-Shirt" to the cart
+        Then my cart taxes should be "€23.00"
+        And my cart total should be "€123.00"
```
I think it's more natural to swap these two:
```
Then my cart total should be ...
And my cart taxes should be
```

WDYT?

pjedrzejewski replied:
Remember to follow:

GIVEN -> WHEN -> THEN

This should be:

```
When I add product "PHP T-Shirt" to the cart
Then my cart taxes should be "€23.00"
And my cart total should be "€123.00"
```

Anyway, we are clearly missing some context here, because we don't know anything about the user or the zone he falls into. We probably need to say: ``And default tax zone is "EU"`` in the background.

pjedrzejewski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+@ui-cart
+Feature: Cart with different tax rates for different time zones
+    In order to order products with different tax rate for different zones
```
``In order to buy goods with correct taxes applied``?

Feature benefit cannot be feature itself...

pjedrzejewski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+@ui-cart
+Feature: Cart with different tax rates for different time zones
+    In order to order products with different tax rate for different zones
+    As a Customer
```
We should decide about the roles in the first place:

* Visitor (not logged in)
* Customer (logged in)
* Guest Customer (not logged in, but in checkout)
* Admin

Thoughts?

michalmarcinkowski replied:
:+1: 
 
* Guest Customer (not logged in, but in checkout) - isn't it a Visitor?

pjedrzejewski replied:
Yes, but I think we should differentiate between someone who is just visiting and someone who started a checkout, we have much more info about him etc.

pjedrzejewski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+Feature: Cart with different tax rates for different time zones
+    In order to order products with different tax rate for different zones
+    As a Customer
+    I want to be aware of taxes applied on my order
```
I want to have correct taxes applied to my order

pjedrzejewski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+    I want to be aware of taxes applied on my order
+
+    Background:
+        Given that store is operating on the France channel
```
I'd change this step to:

``Give the store is operating on a single "France" channel``

michalmarcinkowski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        Then my cart taxes should be "€28.00"
+        And my cart total should be "€228.00"
+
+    Scenario: Proper taxes with rates-zones chaos after addressing
```
`with rates-zones chaos` - not easy to understand

pjedrzejewski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+    Background:
+        Given that store is operating on the France channel
+        And default currency is "EUR"
+        And store has "EU VAT" tax rate of 23% for "Clothes" in "EU" zone
```
``in`` vs. ``within``?

pjedrzejewski replied:
``And store is charging "EU VAT" tax of 23% for "Clothes" within "EU" zone

Zales0123 replied:
I'm :+1: with ``within``

michalmarcinkowski commented on features/shop/cart_with_multiple_tax_rates.feature:
```diff
@@ -0,0 +1,30 @@
+@ui-cart
+Feature: Cart with multiple tax rates
+    In order to order products with different tax rate
```
`order products` -> `buy products` ?

pjedrzejewski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        Given that store is operating on the France channel
+        And default currency is "EUR"
+        And store has "EU VAT" tax rate of 23% for "Clothes" in "EU" zone
+        And store has "No tax" tax rate of 0% for "Clothes" in "Rest of World" zone
```
I think we should write this as:

``And store is charging no tax for "Clothes" for the rest of the world`` (we will have a special CATCH_ALL zone type available to avoid the need of adding all countries to "Rest of World" zone.

pjedrzejewski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        And there is user "john@example.com" identified by "password123"
+        And catalog has a product "PHP T-Shirt" priced at €100.00
+        And "PHP T-Shirt" tax category is "Clothes"
+        And catalog has a product "Symfony Mug" priced at €50.00
```
Maybe we should avoid the word ``catalog`` here and just say ``And store has a product...```, opinions?

pjedrzejewski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        And "PHP T-Shirt" tax category is "Clothes"
+        And catalog has a product "Symfony Mug" priced at €50.00
+        And "Symfony T-Shirt" tax category is "Mugs"
+        And store has free shipping method
```
``And store ships everything for free`` -> This means that there is only free shipping method etc.

pjedrzejewski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        And my cart total should be "€123.00"
+
+    Scenario: Proper taxes after specifying address
+        Given I added product "PHP T-Shirt" to the cart
```
This sounds like an action, so I'd put it in WHENS, unless we change it to ``Given I have product ... in the cart`` but I am not sure about this.

pjedrzejewski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+
+    Scenario: Proper taxes for multiple products after specifying address
+        Given I added 3 products "PHP T-Shirt" to the cart
+        And I proceed selecting "Offline" payment method and "Uzbekistan" as address country
```
When should be here and +1 for ``shipping country`` instead of address country.

pjedrzejewski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        Then my cart taxes should be "€0.00"
+        And my cart total should be "€300.00"
+
+    Scenario: Proper taxes with rates-zones chaos before addressing
```
``chaos`` What?

Zales0123 replied:
This is how it works if you think "I have no idea for this scenario name, just write something stupid for now and fix it in a minutes...". Totally unprofessional, sorry for that :(

pjedrzejewski commented:
I left bunch of comments but I stopped to avoid creating a mess in the diff comments section, so please apply these and other suggestion and we will iterate again. Not an easy task but I think we are very close! :+1:

Zales0123 commented:
@pjedrzejewski @michalmarcinkowski @peteward @pamil Ok :smile: 
I've implemented part of your comments, another one will be soon. Of course I could've missed something, co feel free to point them again.
Some thoughts:
1. I fixed [this](https://github.com/Sylius/Sylius/pull/3941#issuecomment-173846703) comment by specifying tax category target (``product "PHP T-Shirt" belongs to...``/``shipping method "DHL" belongs to..``), I think it's clear enough
2. I think we should stay with quotation marks for prices (``€10.00`` -> ``"€10.00"``) ;)
3. :+1: for avoiding "catalog" word, I just cannot decide which one is better: ``And there is a product "PHP T-Shirt" priced at...`` or ``And store has a product "PHP T-Shirt" priced at...``, WDYT?

michalmarcinkowski commented:
@peteward we definitely are going to have some scenarios for rounding the tax at item / unit level. 

@Zales0123 as for 3. IMO `And store has a product...`

pjedrzejewski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        And I am logged in customer
+
+    Scenario: Displaying correct tax before specifying shipping address
+        When I added product "PHP T-Shirt" to the cart
```
Added suggests it is a pre condition, I think it should be action (When) -> ``add product``

michalmarcinkowski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+    Scenario: Displaying correct tax before specifying shipping address
+        When I added product "PHP T-Shirt" to the cart
+        Then my cart taxes should be "€23.00"
+        And my cart total should be "€123.00"
```
IMO swapping these two will be more natural.

michalmarcinkowski commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        And my cart total should be "€123.00"
+
+    Scenario: Displaying correct tax after specifying shipping address
+        When I added product "PHP T-Shirt" to the cart
```
```
Given I added ...
When I proceed ...
```

lchrusciel commented on features/shop/cart_with_different_taxes_for_different_zones.feature:
```diff
@@ -0,0 +1,50 @@
+        And product "PHP T-Shirt" belongs to "Clothes" tax category
+        And store has a product "Symfony Mug" priced at "€50.00"
+        And product "Symfony Mug" belongs to "Mugs" tax category
+        And store ships everything for free
```
Am I right, that shipment cost is calculated based on zone, not on item? Shouldn't we change this step to _And store ships **everywhere** for free_?

michalmarcinkowski replied:
Actually we ship only to the countries/zones we specified not everywhere in the world, so **everything** is probably more accurate than **everywhere**, WDYT?

lchrusciel replied:
I still found **everywhere** much more accurate. Firstly, as I wrote before, because shipping cost is related to zone, not to item. Secondly, because in this scenario we are delivering items everywhere in the world, actually. But to me more precise, WDYT about `And all store shipments are free of charge`?

michalmarcinkowski replied:
1. Actually the shipping cost can be per item, there are e.g. per item and volume based calculators :)
2. The step is implemented in a way that right now it takes the 'current' zone, so the shipping method is available in one zone, not everywhere.

The most appropriate step will be:
 `And the store ships everything for free everywhere` - this step could get all zones and create free shipping method for every available zone.

pjedrzejewski replied:
``And the store ships everything for free to every possible location``

Still not perfect, because it is not really everywhere, it is only to zones defined, which can be 2 countries. :D

michalmarcinkowski replied:
@pjedrzejewski looks like the most self describing :+1:

michalmarcinkowski commented:
@Zales0123 I suppose this can be closed?

Zales0123 commented:
@michalmarcinkowski I think yes, as we're heading to the end of this scenarios implementation ;)
